File: Algorithm/Other/at coder/Other/B43/B43-gpt.js
// 以下は、Node.js (v18.16.1) + `fs` モジュールを使って、**各生徒の最終的な正解数**を効率的に計算・出力する JavaScript プログラムです。

// * **時間計算量：O(M + N)**（M個の誤答を処理＋N人分の正解数の計算）
// * **空間計算量：O(N)**（各生徒の誤答回数を記録）

// ---

// ### ✅ 解法のポイント

// * 各問題は **N-1人が正解** → 全体の正解数は `M` 問中、自分が間違えた数を引いた `M - 間違えた数`。
// * そこで各生徒について「間違えた回数」をカウントしておけば、全員に対して `M - 間違えた回数` を出力すればよい。

// ---

// ### ✅ 実装コード

// ```javascript
const fs = require('fs');

/**
 * 生徒の正解数を計算する関数
 * @param {number} N - 生徒の人数
 * @param {number} M - 出題された問題数
 * @param {number[]} A - 各問題で間違えた生徒番号の配列（1-indexed）
 * @returns {number[]} - 各生徒の正解数（index: 0-based, 値: 正解数）
 */
function calculateCorrectAnswers(N, M, A) {
    const wrongCounts = new Array(N).fill(0);

    for (let i = 0; i < M; i++) {
        const student = A[i] - 1; // 1-indexed を 0-indexed に変換
        wrongCounts[student]++;
    }

    // 正解数 = M - 間違えた回数
    return wrongCounts.map(wrong => M - wrong);
}

// ===== 標準入力処理 =====
function main() {
    const input = fs.readFileSync('/dev/stdin', 'utf8').trim().split(/\s+/);
    const N = parseInt(input[0]);
    const M = parseInt(input[1]);
    const A = input.slice(2).map(Number);

    const result = calculateCorrectAnswers(N, M, A);
    console.log(result.join('\n'));
}

main();

// ### ✅ 実行方法

// 次のように実行します（`input.txt` に入力を保存しておく場合）：

// ```bash
// node solution.js < input.txt
// ```

// ---

// ### ✅ 入力例

// ```
// 4 6
// 1 4 1 4 2 1
// ```

// ### ✅ 出力例

// ```
// 3
// 5
// 6
// 4
// ```

// ---

// 必要に応じて、プロファイリング (`console.time()` / `process.memoryUsage()` など) も追加可能です。必要でしたら補足します。

File: Algorithm/Other/at coder/Other/B43/B43-gpt.py
# 以下は、**Python (CPython 3.11.4)** を用いて、**型注釈を明示**し、**Pylanceの型検査に準拠**しつつ、**処理時間とメモリ効率を意識**したクイズ大会の正解数計算プログラムです。

# ---

# ## ✅ 解法のポイント

# * `M` 問全てに対して、各生徒の **誤答数** を `O(1)` で記録。
# * 最終的に、各生徒の **正解数 = M - 誤答数** を出力。

# ---

# ## ✅ Python コード（型注釈 + Pylance対応）

# ```python
from typing import List
import sys

# 高速入力用（制約対策）
input = sys.stdin.readline

def calculate_correct_answers(N: int, M: int, A: List[int]) -> List[int]:
    """
    クイズ大会の正解数を計算する

    Parameters:
    - N (int): 生徒の人数 (1 <= N <= 200_000)
    - M (int): 問題数 (1 <= M <= 200_000)
    - A (List[int]): 各問題で誤答した生徒番号（1-indexed, 長さ M）

    Returns:
    - List[int]: 各生徒の正解数（index: 0-based, 値: M - 間違えた回数）
    """
    wrong_counts: List[int] = [0] * N  # 各生徒の誤答数

    for a in A:
        wrong_counts[a - 1] += 1  # 1-indexed → 0-indexed

    # 正解数 = M - 間違えた数
    return [M - wrong for wrong in wrong_counts]

def main() -> None:
    """
    標準入力から N, M, A を読み込み、各生徒の正解数を出力する
    """
    N_str: str = sys.stdin.readline()
    while N_str.strip() == "":
        N_str = sys.stdin.readline()
    N, M = map(int, N_str.strip().split())

    A: List[int] = []
    while len(A) < M:
        line = sys.stdin.readline()
        if not line:
            break
        A.extend(map(int, line.strip().split()))

    result: List[int] = calculate_correct_answers(N, M, A)
    print('\n'.join(map(str, result)))

if __name__ == "__main__":
    main()
# ```

# ---

# ## ✅ 実行例

# ### 入力例（`input.txt`）

# ```
# 4 6
# 1 4 1 4 2 1
# ```

# ### 実行方法

# ```bash
# python3 main.py < input.txt
# ```

# ### 出力

# ```
# 3
# 5
# 6
# 4
# ```

# ---

# ## ✅ 処理時間・メモリ分析

# | 処理        | 時間計算量        | 空間計算量    |
# | --------- | ------------ | -------- |
# | 誤答数の記録    | O(M)         | O(N)     |
# | 正解数の計算    | O(N)         | O(N)     |
# | 出力（print） | O(N)         | O(1)     |
# | **合計**    | **O(N + M)** | **O(N)** |

# * `N, M ≤ 200000` に対しても高速に処理可能。
# * メモリ使用量（概算）：

#   * `wrong_counts`: `4 * N bytes ≈ 0.8MB`（N=200000の場合）

# ---

# ## ✅ Pylance 型検査ポイント

# * `List[int]` 明示により `list` の暗黙型エラー回避。
# * 関数引数・返却値に全て型注釈を付与。
# * `input()` と `sys.stdin.readline()` の混用による `Optional[str]` 混入も排除済。

# ---

# 必要に応じて、`time.perf_counter()` や `tracemalloc` を使って、実際の処理時間・メモリ計測コードも追加できます。ご希望があればお知らせください。
